github-review.el

```diff
@@ -484,48 +476,39 @@ See ‘github-review-start’ for more information"
-                          (chain (github-review-a-get v 'chain)))
-                     ;; And top level comments if there is any
-                     (when (and github-review-fetch-top-level-and-review-comments
-                                (> comms 0))
-                       (setq chain (cons `((function . github-review-get-issue-comments)
-                                           (key . top-level-comments))
-                                         chain)))
-                     ;; And code review if there is any
-                     (when (and github-review-fetch-top-level-and-review-comments
-                                (> review_comments 0))
-                       (setq chain (cons `((function . github-review-get-reviews)
-                                           (key . reviews))
-                                         chain)))
-                     (-> ctx
-                         (github-review-a-assoc 'object v)
-                         (github-review-a-assoc 'chain chain))))))))
-
-
 (defun github-review-start-internal (pr-alist)
   "Start review given PR URL given PR-ALIST."
-    (github-review-chain-calls
-     pr-alist
-     ;; Callback when done
-     (lambda (ctx)
-       (github-review-save-diff
-        pr-alist
-        (github-review-format-diff ctx)))
+  (deferred:$
+    (deferred:parallel
+      ;; Get the diff
+      (lambda () (github-review-get-pr-deferred pr-alist t))
+      ;; And the PR object
+      (lambda () (github-review-get-pr-deferred pr-alist nil))
+      (when github-review-fetch-top-level-and-review-comments
+        ;; And the top level comments
+        (lambda () (github-review-get-issue-comments-deferred pr-alist)))
+      (when github-review-fetch-top-level-and-review-comments
+        ;; And the previous reviews
+        (lambda () (github-review-get-reviews-deferred pr-alist))))
+    (deferred:nextc it
+      (lambda (x)
+        (let* ((diff (-> x (elt 0)))
+              (pr-object (-> x (elt 1)))
+              (comms (github-review-a-get pr-object 'comments))
+              (review_comments (github-review-a-get pr-object 'review_comments))
+              (issues-comments (when (and (> comms 0) github-review-fetch-top-level-and-review-comments) (-> x (elt 2))))
+              (reviews (when (and (> review_comments 0) github-review-fetch-top-level-and-review-comments) github-review-fetch-top-level-and-review-comments (-> x (elt 3)))))
+           (github-review-save-diff
+            pr-alist
+            (github-review-format-diff (-> (github-review-a-empty)
+                       (github-review-a-assoc 'diff diff)
+                       (github-review-a-assoc 'object pr-object)
+                       (github-review-a-assoc 'top-level-comments issues-comments)
+                       (github-review-a-assoc 'reviews reviews)))))))
+    (deferred:error it
+    (lambda (err)
+      (message "Got an error from the GitHub API!")))))
 
```

Comment:
Could we maybe output the error as well?

Reply:
Yep will add that